src/modules/TextMessageModule.cpp

```diff
@@ -14,9 +17,8 @@ ProcessMessage TextMessageModule::handleReceived(const meshtastic_MeshPacket &mp
 #endif
 
     // We only store/display messages destined for us.
-    // Keep a copy of the most recent text message.
-    devicestate.rx_text_message = mp;
-    devicestate.has_rx_text_message = true;
+
+    nodeDB->saveMessageToDisk(mp);
```

Comment:
I don’t think it’s a good idea to save to flash every time it receives a message, and especially not if the CannedMessageModule is not enabled (or it doesn’t even have a screen). Saving to flash will wear it out, and it can take a long time (especially on nRF52 as @prokrypt recently found), so it even affects packet reception and routing.

Also it uses precious RAM for devices without screen, which should be avoided in my opinion.

Reply:
Throttling this like we're doing for nodedb updates when we have an updated NodeInfo perhaps would be helpful

Reply:
I would add it only to the RAM cache when the TextMessageModule processes it and save it to flash at a different place. E.g. in `NodeDB::loadMessage()` (only when there's a new message) where you also potentially remove messages from the cache and when rebooting/shutting down.

Reply:
This new feature does not seem practical for use on nRF52 devices if it takes a long time to save messages. On top of that there is no way of checking how much free RAM is left as far as I can tell. I will make it fall back on the old way of displaying messages for nRF52 devices, devices with the router role, devices without a screen, and devices where the canned message module is disabled.

As for the flash wearing out, where can I find information on the amount of reads/writes that can be done? It would seem to me that the same amount of reads and writes are occurring after these changes because every message received was being saved to disk before I changed anything. I don't think it is a good idea to save messages in batches because power loss would cause messages to be lost.

Reply:
For ESP32 it's around [100k](https://docs.espressif.com/projects/esp-faq/en/latest/software-framework/storage/nvs.html#if-data-needs-to-be-stored-or-updated-to-flash-every-minute-can-esp32-nvs-meet-this-requirement) writes, for nRF52 more like [10k](https://devzone.nordicsemi.com/f/nordic-q-a/15408/nrf52-write-erase-flash-cycles-confusion#:~:text=Flash%20write%2Ferase%20cycles%3A%2010,a%20block%20between%20erase%3A%20181). However, even for ESP32 saving to flash takes a considerable amount of time.

As you can see here, previously it would only set the last received message, but it only gets written to flash when the devicestate is written to (e.g. with `saveDeviceStateToDisk()`). So, yes, if power is lost in between, it might be some messages are missed, but I think that's not a big deal.

I think people would find it a shame if it's completely removed for nRF52, especially for @tropho23's Meshenger, who originally opened feature requests #1985 and #2590.

`MemGet::getFreeHeap()` is already implemented for nRF52 as well:
https://github.com/meshtastic/firmware/blob/859fd7c25110c038bcb41aa7ef01d6b603b8ab9b/src/memGet.cpp#L23-L25

Reply:
From firsthand experience, with the cache being implemented as it currently is, I haven't noticed the time it takes to save messages.

For nRF52 devices, their future is in external storage devices. There is no support for that but non-nRF52 users will really benefit from this. Especially given that this update will only be enabled for the T-Deck and will be enabled for more devices in the future.

Reply:
I have not seen any impact to the speed either and I am using this on heltec v3, wireless paper and wireless tracker.

Reply:
I'm not really concerned about "speed" in a way that it slows messaging down (indeed you wouldn't really notice an additional 500ms delay). However, the action is not asynchronous, so I'm wondering if you miss packets when you receive several short text messages after each other, just like happened before e-ink updates were implemented asynchronously.

Reply:
Using a Heltec Wireless Paper, I added `millis()` around `nodeDB→saveMessageToDisk(mp);` and it varied from 187-736ms; it’s mostly around 350ms. Considering a message on LongFast usually takes around 500ms, this is a considerable amount of time. The radio chip has place in the buffer for only 1 packet and it's shared between transmit and receive.

I also did a test by sending 4 times 5 text messages with 1 character quickly after each other. Two of them didn’t get acknowledged, and the packets were often received out of order, meaning that retransmissions were needed (which is only done for the 0-hop, so for further hops it would be lost).

With the same test on current master, all 20 packets got acknowledged.